scipy/stats/_qmc_cy.pyx

```diff
@@ -0,0 +1,326 @@
+# cython: language_level=3
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+
+# distutils: language = c++
+
+import numpy as np
+cimport numpy as np
+from libc.math cimport fabs, sqrt, pow
+
+np.import_array()
+
+cdef extern from "<thread>" namespace "std" nogil:
+    cdef cppclass thread:
+        thread()
+        void thread[A, B, C, D, E, F](A, B, C, D, E, F)
+        void join()
+
+cdef extern from "<mutex>" namespace "std" nogil:
+    cdef cppclass mutex:
+        void lock()
+        void unlock()
+
+cdef extern from "<functional>" namespace "std" nogil:
+    cdef cppclass reference_wrapper[T]:
+        pass
+    cdef reference_wrapper[T] ref[T](T&)
+
+from libcpp.vector cimport vector
+
+
+cdef mutex threaded_sum_mutex
+
+def _cy_wrapper_centered_discrepancy(double[:, ::1] sample, bint iterative,
+                                    workers):
+        return centered_discrepancy(sample, iterative, workers)
+
+
+def _cy_wrapper_wrap_around_discrepancy(double[:, ::1] sample,
+                                       bint iterative, workers):
+        return wrap_around_discrepancy(sample, iterative, workers)
+
+
+def _cy_wrapper_mixture_discrepancy(double[:, ::1] sample,
+                                   bint iterative, workers):
+        return mixture_discrepancy(sample, iterative, workers)
+
+
+def _cy_wrapper_l2_star_discrepancy(double[:, ::1] sample,
+                                   bint iterative, workers):
+        return l2_star_discrepancy(sample, iterative, workers)
+
+
+cdef double centered_discrepancy(double[:, ::1] sample_view,
+                             bint iterative, unsigned int workers) nogil:
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0
+        double prod, disc1 = 0
+
+    for i in range(n):
+        prod = 1
+        for j in range(d):
+                prod *= (
+                        1 + 0.5 * fabs(sample_view[i, j] - 0.5) - 0.5 * fabs(
+                        sample_view[i, j] - 0.5) ** 2
+                )
+        disc1 += prod
```

Comment:
Just to understand. I suppose that this is not in the `threaded_loops` due to the overhead which would compensate any gain?

Reply:
Yes. This first loop has a `O(nd)` complexity. The second loop has a `O(n²d)` complexity. (The constant should be near the same). I think the the gain for the first loop is very small, introduce a overhead, and a maintenance complexity.

Reply:
Indeed, just wanted to have some confirmation about the overhead. Thanks.